File: workers/booking/modules/limousine_p2d4c1_branding_hero_offers.test.mjs
import test from "node:test";
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

import {
  applyPublicLimousineHeroFields,
  normalizeLimousineHero,
  normalizeLimousinePricingSection,
} from "./limousine_pricing_resolver.mjs";
import {
  buildSafePublicLimousineOffers,
  normalizeLimousineOffer,
} from "./limousine_offers.mjs";
import { buildLimousineNearbyCardProjection } from "./limousine_discovery_preview.mjs";

const here = dirname(fileURLToPath(import.meta.url));
const worker = readFileSync(join(here, "../fluxidi_booking_worker.js"), "utf8");

const CLASS_IDS = ["party_stretch", "suv_stretch"];
const FLEET = [
  {
    id: "vh_party",
    is_active: true,
    service_category: "limousine",
    service_class_id: "party_stretch",
  },
  {
    id: "vh_hummer",
    is_active: true,
    service_category: "limousine",
    service_class_id: "suv_stretch",
  },
  {
    id: "vh_hidden",
    is_active: true,
    service_category: "",
    service_class_id: "",
  },
];

function offer(overrides = {}) {
  return {
    offer_id: "off_1",
    enabled: true,
    published: true,
    target_type: "vehicle",
    vehicle_id: "vh_party",
    vehicle_ids: ["vh_party"],
    applies_to_all_selected_vehicles: false,
    price_presentation: "from_price",
    display_amount_cents: 25000,
    currency: "EUR",
    title: { nl: "Party", en: "Party", fr: "", es: "" },
    ...overrides,
  };
}

test("hero persists on the pricing section and never copies taxi cover", () => {
  const section = normalizeLimousinePricingSection({
    enabled: true,
    currency: "EUR",
    limousine_hero: {
      photo_url: "https://cdn.example/limousine-hero.jpg",
      source_kind: "upload",
      alignment: "top",
      source_revision: 3,
    },
  });
  assert.equal(section.limousine_hero.photo_url, "https://cdn.example/limousine-hero.jpg");
  assert.equal(section.limousine_hero.source_kind, "upload");
  assert.equal(section.limousine_hero.alignment, "top");
  const again = normalizeLimousineHero(section);
  assert.equal(again.photo_url, section.limousine_hero.photo_url);
  const stamped = applyPublicLimousineHeroFields(
    {
      hero_photo_url: "https://cdn.example/taxi-cover.jpg",
      media: { hero_photo_url: "https://cdn.example/taxi-cover.jpg" },
    },
    section,
  );
  assert.equal(stamped.limousine_hero_url, "https://cdn.example/limousine-hero.jpg");
  assert.equal(stamped.hero_photo_url, "https://cdn.example/taxi-cover.jpg");
  const cleared = applyPublicLimousineHeroFields(stamped, { limousine_hero: {} });
  assert.equal(cleared.limousine_hero_url, undefined);
});

test("legacy unbound offers become all-selected; vehicle bindings stay distinct", () => {
  const legacy = normalizeLimousineOffer({
    offer_id: "off_legacy",
    enabled: true,
    published: true,
    target_type: "service_class",
    service_class_id: "party_stretch",
    price_presentation: "quote_required",
    currency: "EUR",
  });
  assert.equal(legacy.applies_to_all_selected_vehicles, true);
  assert.deepEqual(legacy.vehicle_ids, []);

  const party = normalizeLimousineOffer(offer());
  const hummer = normalizeLimousineOffer(
    offer({
      offer_id: "off_hummer",
      vehicle_id: "vh_hummer",
      vehicle_ids: ["vh_hummer"],
      price_presentation: "exact_fixed",
      display_amount_cents: 65000,
    }),
  );
  const publicOffers = buildSafePublicLimousineOffers([party, hummer], {
    eligible: true,
    knownVehicles: FLEET,
    knownClassIds: CLASS_IDS,
    readiness: true,
  });
  const partyPublic = publicOffers.find((item) => item.offer_id === "off_1");
  const hummerPublic = publicOffers.find((item) => item.offer_id === "off_hummer");
  assert.deepEqual(partyPublic.vehicle_ids, ["vh_party"]);
  assert.deepEqual(hummerPublic.vehicle_ids, ["vh_hummer"]);
  assert.equal(partyPublic.display_amount_cents, 25000);
  assert.equal(hummerPublic.display_amount_cents, 65000);
});

test("unpublished linked vehicle is omitted without deleting the offer", () => {
  const publicOffers = buildSafePublicLimousineOffers(
    [
      offer({
        vehicle_ids: ["vh_party", "vh_hidden"],
        vehicle_id: "vh_party",
      }),
    ],
    {
      eligible: true,
      knownVehicles: FLEET,
      knownClassIds: CLASS_IDS,
      readiness: true,
    },
  );
  assert.equal(publicOffers.length, 1);
  assert.deepEqual(publicOffers[0].vehicle_ids, ["vh_party"]);
  assert.equal(publicOffers[0].vehicle_ids.includes("vh_hidden"), false);
});

test("shared and all-selected offers publish for every classified limousine", () => {
  const shared = buildSafePublicLimousineOffers(
    [
      offer({
        offer_id: "off_both",
        vehicle_ids: ["vh_party", "vh_hummer"],
        vehicle_id: "vh_party",
        price_presentation: "quote_required",
        display_amount_cents: null,
      }),
    ],
    {
      eligible: true,
      knownVehicles: FLEET,
      knownClassIds: CLASS_IDS,
      readiness: true,
    },
  );
  assert.deepEqual(shared[0].vehicle_ids, ["vh_party", "vh_hummer"]);

  const all = buildSafePublicLimousineOffers(
    [
      offer({
        offer_id: "off_all",
        applies_to_all_selected_vehicles: true,
        vehicle_ids: [],
        vehicle_id: "",
        target_type: "service_class",
        service_class_id: "party_stretch",
        price_presentation: "quote_required",
        display_amount_cents: null,
      }),
    ],
    {
      eligible: true,
      knownVehicles: FLEET,
      knownClassIds: CLASS_IDS,
      readiness: true,
    },
  );
  assert.equal(all[0].applies_to_all_selected_vehicles, true);
});

test("nearby summary prefers featured then lowest from-price and never labels a package as from", () => {
  const card = buildLimousineNearbyCardProjection({
    partner_id: "limo_1",
    company_name: "Maison",
    is_active: true,
    bookable: true,
    profile_enabled: true,
    published_at: "2026-08-17T10:00:00Z",
    subscription_status: "active",
    limousine_entitled: true,
    limousine_available: true,
    limousine_service_enabled: true,
    services: ["limousine"],
    logo_url: "https://cdn.example/logo.png",
    vehicles: [
      {
        service_category: "limousine",
        service_class: "party_stretch",
        photo_url: "https://cdn.example/party.jpg",
        pax: 8,
        is_active: true,
      },
    ],
    limousine_offers: [
      {
        offer_id: "off_pkg",
        published: true,
        enabled: true,
        price_presentation: "exact_fixed",
        display_amount_cents: 9000,
        currency: "EUR",
        hourly: { enabled: true, package_amount_cents: 65000, package_duration_minutes: 240 },
      },
      {
        offer_id: "off_from",
        published: true,
        enabled: true,
        price_presentation: "from_price",
        display_amount_cents: 25000,
        currency: "EUR",
        featured: true,
      },
    ],
  });
  assert.equal(card.logo_url, "https://cdn.example/logo.png");
  assert.equal(card.limousine_price_presentation, "from_price");
  assert.equal(card.display_amount_cents, 25000);
});

test("worker preserves public hero fields and limousine vehicle_id", () => {
  assert.match(worker, /limousine_hero_url/);
  assert.match(worker, /_applyPublicLimousineHeroFields/);
  assert.match(worker, /serviceCategory === "limousine" && publicVehicleId/);
  assert.match(worker, /applies_to_all_selected_vehicles/);
  assert.doesNotMatch(worker, /hero_photo_url:\s*hero\.photo_url/);
});

File: test/limousine/limousine_p2d4c1_branding_hero_offers_test.dart
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:fluxidi_tracking/app_config.dart';
import 'package:fluxidi_tracking/app_strings.dart';
import 'package:fluxidi_tracking/customer_theme_store.dart';
import 'package:fluxidi_tracking/limousine/limousine_brand_logo.dart';
import 'package:fluxidi_tracking/limousine/limousine_customer_discovery.dart';
import 'package:fluxidi_tracking/limousine/limousine_hero_contract.dart';
import 'package:fluxidi_tracking/limousine/limousine_offer_binding.dart';
import 'package:fluxidi_tracking/limousine/limousine_offers.dart';
import 'package:fluxidi_tracking/customer_theme_palette.dart';
import 'package:fluxidi_tracking/limousine/limousine_provider_showroom.dart';
import 'package:fluxidi_tracking/limousine/limousine_provider_showroom_page.dart';
import 'package:fluxidi_tracking/limousine/limousine_public_profile_page.dart';
import 'package:fluxidi_tracking/limousine/limousine_vehicle_detail_page.dart';
import 'package:fluxidi_tracking/limousine/limousine_vehicle_media.dart';

Map<String, dynamic> _vehicle({
  required String id,
  required String name,
  required String photo,
  String serviceClass = 'party_stretch',
}) {
  return <String, dynamic>{
    'vehicle_id': id,
    'name': name,
    'service_category': 'limousine',
    'service_class': serviceClass,
    'pax': 8,
    'luggage': 4,
    'photo_url': photo,
    'primary_photo_url': photo,
    'is_active': true,
  };
}

Map<String, dynamic> _offer({
  required String id,
  required String presentation,
  List<String> vehicleIds = const <String>[],
  bool appliesToAll = false,
  bool featured = false,
  int? amount,
  Map<String, dynamic>? hourly,
  String title = 'Aanbod',
}) {
  return <String, dynamic>{
    'offer_id': id,
    'enabled': true,
    'published': true,
    'target_type': appliesToAll
        ? LimousineOfferTarget.serviceClass
        : LimousineOfferTarget.vehicle,
    'service_class_id': 'party_stretch',
    'vehicle_id': vehicleIds.isEmpty ? '' : vehicleIds.first,
    'vehicle_ids': vehicleIds,
    'applies_to_all_selected_vehicles': appliesToAll,
    'featured': featured,
    'sort_order': 0,
    'price_presentation': presentation,
    'display_amount_cents': amount,
    'currency': 'EUR',
    'title': <String, String>{
      'nl': title,
      'en': title,
      'fr': title,
      'es': title,
    },
    'description': <String, String>{
      'nl': 'Beschrijving $title',
      'en': 'Description $title',
      'fr': 'Description $title',
      'es': 'Descripción $title',
    },
    if (hourly != null) 'hourly': hourly,
  };
}

Map<String, dynamic> _profile({
  String? hero,
  List<Map<String, dynamic>>? vehicles,
  List<Map<String, dynamic>>? offers,
}) {
  return <String, dynamic>{
    'partner_id': 'limo_1',
    'company_name': 'Maison Noire',
    'logo_url': 'https://cdn.example/logo.png',
    'hero_photo_url': 'https://cdn.example/taxi-cover.jpg',
    'limousine_available': true,
    if (hero != null) 'limousine_hero_url': hero,
    'vehicles': vehicles ??
        <Map<String, dynamic>>[
          _vehicle(
            id: 'vh_party',
            name: 'Party Limo',
            photo: 'https://cdn.example/party.jpg',
          ),
          _vehicle(
            id: 'vh_hummer',
            name: 'Hummer white',
            photo: 'https://cdn.example/hummer.jpg',
            serviceClass: 'suv_stretch',
          ),
        ],
    'limousine_offers': offers ??
        <Map<String, dynamic>>[
          _offer(
            id: 'off_party',
            presentation: LimousinePricePresentation.fromPrice,
            vehicleIds: const <String>['vh_party'],
            amount: 25000,
            title: 'Party vanaf',
          ),
          _offer(
            id: 'off_hummer',
            presentation: LimousinePricePresentation.exactFixed,
            vehicleIds: const <String>['vh_hummer'],
            amount: 65000,
            title: 'Hummer avondrit',
          ),
        ],
  };
}

Widget _app(Widget child) {
  return MaterialApp(
    home: MediaQuery(
      data: const MediaQueryData(size: Size(800, 1280)),
      child: child,
    ),
  );
}

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  tearDown(() {
    appLanguageNotifier.value = AppLanguage.nl;
    customerThemeNotifier.value = CustomerThemeVariant.premiumLight;
  });

  test('1-4 explicit limousine hero persists and ignores fleet/taxi cover', () {
    const explicit = 'https://cdn.example/limousine-hero.jpg';
    final first = resolveLimousineHero(
      source: <String, dynamic>{
        'limousine_hero_url': explicit,
        'limousine_hero_source': 'upload',
        'hero_photo_url': 'https://cdn.example/taxi-cover.jpg',
      },
      fallbackVehiclePhotoUrls: const <String>[
        'https://cdn.example/party.jpg',
        'https://cdn.example/hummer.jpg',
      ],
    );
    expect(first.photoUrl, explicit);
    expect(first.explicit, isTrue);
    final reordered = resolveLimousineHero(
      source: <String, dynamic>{'limousine_hero_url': explicit},
      fallbackVehiclePhotoUrls: const <String>[
        'https://cdn.example/hummer.jpg',
        'https://cdn.example/party.jpg',
      ],
    );
    expect(reordered.photoUrl, explicit);
    final fallback = resolveLimousineHero(
      source: <String, dynamic>{
        'hero_photo_url': 'https://cdn.example/taxi-cover.jpg',
      },
      fallbackVehiclePhotoUrls: const <String>['https://cdn.example/party.jpg'],
    );
    expect(fallback.photoUrl, 'https://cdn.example/party.jpg');
    expect(fallback.explicit, isFalse);
    expect(fallback.photoUrl.contains('taxi-cover'), isFalse);
    final empty = resolveLimousineHero(
      source: <String, dynamic>{
        'hero_photo_url': 'https://cdn.example/taxi-cover.jpg',
      },
    );
    expect(empty.photoUrl, isEmpty);
    final showroom = buildLimousineProviderShowroomData(
      profile: _profile(hero: explicit),
    );
    expect(showroom.heroPhotoUrl, explicit);
    expect(showroom.heroIsExplicit, isTrue);
    expect(showroom.heroPhotoUrl.contains('taxi-cover'), isFalse);
    expect(
      buildLimousineProviderShowroomData(profile: _profile()).heroPhotoUrl,
      'https://cdn.example/party.jpg',
    );
  });

  test('discovery keeps vehicle primary photo, not company hero', () {
    final card = tryParseLimousineDiscoveryCard(<String, dynamic>{
      'partner_id': 'limo_1',
      'company_name': 'Maison Noire',
      'is_active': true,
      'profile_enabled': true,
      'limousine_available': true,
      'limousine_service_enabled': true,
      'logo_url': 'https://cdn.example/logo.png',
      'limousine_hero_url': 'https://cdn.example/limousine-hero.jpg',
      'hero_photo_url': 'https://cdn.example/taxi-cover.jpg',
      'limousine_vehicles': <Map<String, dynamic>>[
        <String, dynamic>{
          'service_category': 'limousine',
          'service_class_id': 'party_stretch',
          'photo_url': 'https://cdn.example/party.jpg',
        },
      ],
      'limousine_offers': <Map<String, dynamic>>[
        _offer(
          id: 'off_party',
          presentation: LimousinePricePresentation.fromPrice,
          vehicleIds: const <String>['vh_party'],
          amount: 25000,
        ),
      ],
    });
    expect(card?.coverImageUrl, 'https://cdn.example/party.jpg');
    expect(card?.logoUrl, 'https://cdn.example/logo.png');
    expect(card?.price.kind, LimousineDiscoveryPriceKind.fromPrice);
    expect(card?.price.amountCents, 25000);
  });

  testWidgets('5-6 logo plaque is visible and keeps contain aspect ratio', (
    tester,
  ) async {
    final profile = _profile(hero: 'https://cdn.example/limousine-hero.jpg');
    await tester.pumpWidget(
      _app(LimousineProviderShowroomPage(partnerId: 'limo_1', profile: profile)),
    );
    await tester.pump();
    expect(find.byKey(kLimousineBrandLogoPlaqueKey), findsWidgets);
    expect(
      find.byKey(kLimousineBrandLogoImageKey).evaluate().isNotEmpty ||
          find.byKey(kLimousineBrandLogoInitialsKey).evaluate().isNotEmpty,
      isTrue,
    );
    if (find.byKey(kLimousineBrandLogoImageKey).evaluate().isNotEmpty) {
      expect(
        tester.widget<Image>(find.byKey(kLimousineBrandLogoImageKey).first).fit,
        BoxFit.contain,
      );
    }
    await tester.pumpWidget(
      _app(LimousinePublicProfilePage(partnerId: 'limo_1', profile: profile)),
    );
    await tester.pump();
    expect(find.byKey(kLimousineBrandLogoPlaqueKey), findsWidgets);
    final data = buildLimousineProviderShowroomData(profile: profile);
    await tester.pumpWidget(
      _app(
        LimousineVehicleDetailPage(
          vehicle: data.vehicles.first,
          companyName: 'Maison Noire',
          partnerId: 'limo_1',
          logoUrl: data.logoUrl,
        ),
      ),
    );
    await tester.pump();
    expect(find.byKey(kLimousineBrandLogoPlaqueKey), findsWidgets);
    final logo = File('lib/limousine/limousine_brand_logo.dart').readAsStringSync();
    expect(logo.contains('BoxFit.contain'), isTrue);
    expect(logo.contains('BoxFit.fill'), isFalse);
  });

  test('7-12 Party and Hummer keep distinct bindings, shared and all-selected work', () {
    final data = buildLimousineProviderShowroomData(profile: _profile());
    final party = data.vehicles.firstWhere((v) => v.vehicleId == 'vh_party');
    final hummer = data.vehicles.firstWhere((v) => v.vehicleId == 'vh_hummer');
    expect(party.offers.single.offerId, 'off_party');
    expect(hummer.offers.single.offerId, 'off_hummer');
    expect(party.offers.any((o) => o.offerId == 'off_hummer'), isFalse);
    expect(hummer.offers.any((o) => o.offerId == 'off_party'), isFalse);
    expect(
      limousineFormatPublishedOfferPrice(party.offers.single, AppLanguage.nl),
      contains('Vanaf'),
    );
    expect(
      limousineFormatPublishedOfferPrice(hummer.offers.single, AppLanguage.nl),
      isNot(contains('Vanaf')),
    );

    final shared = buildLimousineProviderShowroomData(
      profile: _profile(
        offers: <Map<String, dynamic>>[
          _offer(
            id: 'off_both',
            presentation: LimousinePricePresentation.quoteRequired,
            vehicleIds: const <String>['vh_party', 'vh_hummer'],
            title: 'Beide',
          ),
        ],
      ),
    );
    expect(shared.vehicles.every((v) => v.offers.single.offerId == 'off_both'), isTrue);

    final allSelected = buildLimousineProviderShowroomData(
      profile: _profile(
        offers: <Map<String, dynamic>>[
          _offer(
            id: 'off_all',
            presentation: LimousinePricePresentation.quoteRequired,
            appliesToAll: true,
            title: 'Alle',
          ),
        ],
      ),
    );
    expect(
      allSelected.vehicles.every((v) => v.offers.single.offerId == 'off_all'),
      isTrue,
    );
  });

  test('13 legacy unbound offers stay compatible as all selected', () {
    final legacy = <String, dynamic>{
      'offer_id': 'off_legacy',
      'enabled': true,
      'published': true,
      'target_type': LimousineOfferTarget.serviceClass,
      'service_class_id': 'party_stretch',
      'price_presentation': LimousinePricePresentation.quoteRequired,
      'currency': 'EUR',
      'title': <String, String>{'nl': 'Legacy', 'en': 'Legacy', 'fr': '', 'es': ''},
    };
    final scope = limousineOfferScopeOf(legacy);
    expect(scope.legacyUnbound, isTrue);
    expect(scope.appliesToAllSelected, isTrue);
    final data = buildLimousineProviderShowroomData(
      profile: _profile(offers: <Map<String, dynamic>>[legacy]),
    );
    expect(data.vehicles.every((v) => v.offers.single.offerId == 'off_legacy'), isTrue);
    final saved = limousineWriteOfferScope(
      legacy,
      appliesToAllSelected: true,
      vehicleIds: const <String>[],
    );
    expect(saved['applies_to_all_selected_vehicles'], isTrue);
    expect(limousineOfferScopeOf(saved).legacyUnbound, isFalse);
  });

  test('14-16 detail models stay correctly labelled', () {
    final hourly = _offer(
      id: 'off_hour',
      presentation: LimousinePricePresentation.fromPrice,
      vehicleIds: const <String>['vh_party'],
      amount: 12500,
      hourly: <String, dynamic>{
        'enabled': true,
        'first_hour_cents': 12500,
        'minimum_duration_minutes': 180,
      },
      title: 'Uurhuur',
    );
    final package = _offer(
      id: 'off_pkg',
      presentation: LimousinePricePresentation.exactFixed,
      vehicleIds: const <String>['vh_party'],
      amount: 65000,
      hourly: <String, dynamic>{
        'enabled': true,
        'package_amount_cents': 65000,
        'package_duration_minutes': 240,
      },
      title: 'Trouwarrangement',
    );
    final fixed = _offer(
      id: 'off_fixed',
      presentation: LimousinePricePresentation.exactFixed,
      vehicleIds: const <String>['vh_party'],
      amount: 30000,
      title: 'Avondrit van 3 uur',
    );
    final from = _offer(
      id: 'off_from',
      presentation: LimousinePricePresentation.fromPrice,
      vehicleIds: const <String>['vh_party'],
      amount: 25000,
      title: 'Standaard',
    );
    expect(limousineOfferDisplayKindOf(from), LimousineOfferDisplayKind.fromPrice);
    expect(limousineOfferDisplayKindOf(fixed), LimousineOfferDisplayKind.fixed);
    expect(limousineOfferDisplayKindOf(hourly), LimousineOfferDisplayKind.hourly);
    expect(limousineOfferDisplayKindOf(package), LimousineOfferDisplayKind.package);
    final data = buildLimousineProviderShowroomData(
      profile: _profile(offers: <Map<String, dynamic>>[from, fixed, hourly, package]),
    );
    final party = data.vehicles.firstWhere((v) => v.vehicleId == 'vh_party');
    expect(party.offers, hasLength(4));
    expect(
      limousineFormatPublishedOfferPrice(party.offers.firstWhere((o) => o.offerId == 'off_from'), AppLanguage.nl),
      contains('Vanaf'),
    );
    expect(
      limousineFormatPublishedOfferPrice(party.offers.firstWhere((o) => o.offerId == 'off_fixed'), AppLanguage.nl),
      isNot(contains('Vanaf')),
    );
    expect(
      limousineFormatPublishedOfferPrice(party.offers.firstWhere((o) => o.offerId == 'off_pkg'), AppLanguage.nl),
      allOf(contains('Trouwarrangement'), isNot(contains('Vanaf'))),
    );
    expect(
      limousineFormatPublishedOfferPrice(party.offers.firstWhere((o) => o.offerId == 'off_hour'), AppLanguage.nl),
      contains('per uur'),
    );
  });

  test('17-18 discovery summary and unpublished vehicle do not leak', () {
    final summary = limousineDiscoveryPriceFromOffers(<Map<String, dynamic>>[
      _offer(
        id: 'off_from',
        presentation: LimousinePricePresentation.fromPrice,
        vehicleIds: const <String>['vh_party'],
        amount: 25000,
      ),
      _offer(
        id: 'off_fixed',
        presentation: LimousinePricePresentation.exactFixed,
        vehicleIds: const <String>['vh_hummer'],
        amount: 9000,
      ),
    ]);
    expect(summary.kind, LimousineDiscoveryPriceKind.fromPrice);
    expect(summary.amountCents, 25000);
    final data = buildLimousineProviderShowroomData(
      profile: _profile(
        vehicles: <Map<String, dynamic>>[
          _vehicle(
            id: 'vh_party',
            name: 'Party Limo',
            photo: 'https://cdn.example/party.jpg',
          ),
        ],
        offers: <Map<String, dynamic>>[
          _offer(
            id: 'off_hummer',
            presentation: LimousinePricePresentation.exactFixed,
            vehicleIds: const <String>['vh_hummer'],
            amount: 65000,
            title: 'Hummer only',
          ),
        ],
      ),
    );
    expect(data.vehicles.single.offers, isEmpty);
    expect(
      limousineShowroomVehiclePriceLabel(data.vehicles.single, AppLanguage.nl),
      'Prijs op aanvraag',
    );
  });

  testWidgets('19 gates off create no request or booking', (tester) async {
    var quotes = 0;
    var books = 0;
    final data = buildLimousineProviderShowroomData(profile: _profile());
    await tester.binding.setSurfaceSize(const Size(800, 1280));
    addTearDown(() => tester.binding.setSurfaceSize(null));
    await tester.pumpWidget(
      _app(
        LimousineVehicleDetailPage(
          vehicle: data.vehicles.first,
          companyName: 'Maison Noire',
          partnerId: 'limo_1',
          logoUrl: data.logoUrl,
          onQuote: (_) => quotes++,
          onBook: (_) => books++,
        ),
      ),
    );
    await tester.pump();
    expect(find.byKey(kLimousineDetailPricesSectionKey), findsOneWidget);
    expect(find.text('Prijzen en arrangementen'), findsOneWidget);
    await tester.ensureVisible(find.byKey(kLimousineDetailQuoteCtaKey));
    await tester.tap(find.byKey(kLimousineDetailQuoteCtaKey));
    await tester.pump();
    expect(quotes, 0);
    expect(books, 0);
    expect(find.byKey(kLimousineDetailGateOffBannerKey), findsOneWidget);
  });

  test('20 save/reopen keeps hero, vehicles and offer bindings', () {
    final section = <String, dynamic>{
      'selected_vehicle_ids': <String>['vh_party', 'vh_hummer'],
      'limousine_hero': <String, dynamic>{
        'photo_url': 'https://cdn.example/limousine-hero.jpg',
        'source_kind': 'upload',
        'alignment': 'top',
        'source_revision': 4,
      },
      'offers': <Map<String, dynamic>>[
        limousineWriteOfferScope(
          _offer(
            id: 'off_party',
            presentation: LimousinePricePresentation.fromPrice,
            vehicleIds: const <String>['vh_party'],
            amount: 25000,
          ),
          appliesToAllSelected: false,
          vehicleIds: const <String>['vh_party'],
          featured: true,
          sortOrder: 2,
        ),
      ],
    };
    final hero = limousineHeroFromSection(section);
    expect(hero.photoUrl, 'https://cdn.example/limousine-hero.jpg');
    expect(hero.alignment, 'top');
    expect(hero.sourceRevision, 4);
    expect(hero.explicit, isTrue);
    final again = limousineHeroFromSection(<String, dynamic>{
      'limousine_hero': hero.toSectionJson(),
    });
    expect(again.photoUrl, hero.photoUrl);
    expect(again.alignment, hero.alignment);
    final scope = limousineOfferScopeOf(
      (section['offers'] as List).first as Map<String, dynamic>,
    );
    expect(scope.vehicleIds, <String>['vh_party']);
    expect(scope.featured, isTrue);
    expect(scope.sortOrder, 2);
    expect(scope.appliesToAllSelected, isFalse);
  });

  test('vehicle media stay contain while explicit hero may cover', () {
    expect(
      limousineVehicleMediaUsesContainStrategy(
        minHeight: kLimousineVehiclePhotoMinHeight,
        photoFit: BoxFit.contain,
      ),
      isTrue,
    );
    expect(limousineCompanyInitials('Maison Noire'), 'MN');
    expect(limousineCompanyInitials(''), 'L');
  });
}
